### PR TITLE
feat(tsserver): set `single_file_support` to true

### PR DESCRIPTION
**Pull request by rjooske**
Set `single_file_support` for tsserver to true by default, as typescript-language-server supports single-file mode.

### CODE REVIEW

**Comment by glepnir** [comment]
can you post the relate doc ? I don't find the typescipt-language-server support single file.

**Comment by rjooske** [comment]
Actually, I also have tried to find something that mentions typescript-language-server's single file support, but couldn't find any.
My original idea was that because VSCode uses typescript-language-server and you can edit a single JS, TS, and such files with LSP features on VSCode, there's no reason that the same can't be done on Neovim.
I've been using Neovim + typescript-language-server with `single_file_support` = true, and so far it seems to work exactly like how VSCode works when editing a single JS, TS, etc files.
I'm sorry but I can't provide any other related (and possibly useful) information :(
Hope this helps.

**Comment by glepnir** [comment]
vscode use tsserver not this server.

**Comment by rjooske** [comment]
Oh okay, I think I was very confused about the terminology here.
Let me see if I understand it correctly now:
- tsserver
  - is part of the TypeScript ecosystem
  - is used by VSCode for completion, hover, etc
  - is **the name** of the server config for JS/TS in nvim-lspconfig, but NOT actually used by nvim-lspconfig
- typescript-language-server
  - is a language server that can speak LSP
  - is used by nvim-lspconfig to implement LSP features, under the name "tsserver"

And my assumption is that:
- tsserver
  - supports single file mode because VSCode's behavior suggests so
- typescript-language-server
  - supports single file mode because after setting `single_file_support` to true, completion, hover, etc are working as expected on Neovim

So to summarize, my original reason as to why I thought setting `single_file_support` to true would work turns out to be very much wrong, but the outcome (being able to use LSP features) happens to match my expectation by chance.
Although it doesn't answer the question of whether there is documentation about typescript-language-server's single file support...
All I can do now is say "it works on my machine" ¯\_(ツ)_/¯

**Comment by glepnir** [comment]
nope . the typescript-language-server is wrapped tsserver. tsserver only work in vscode.  about `single_file_mode ` in typescript-language-server.  it's wrapped tsserver so I think it should support `single_file_mode`. but I don't know why does not add this option in default. maybe there has something wrong or something else. maybe you can test all lsp functions with  `single_file_mode` , if both works fine. we can add it into default .

**Comment by rjooske** [comment]
> the typescript-language-server is wrapped tsserver

That makes a lot of sense!

>  you can test all lsp functions with `single_file_mode`

Sorry, I'm not entirely sure about what you mean by "test"...

**Comment by glepnir** [comment]
oh I mean the functions `vim.lsp.buf.*`

**Comment by rjooske** [comment]
Okay, so... are we going to make sure that functions in `vim.lsp.buf.*` don't do anything out of the ordinary with `single_file_support = true` or something? I still don't understand what you want to guarantee by "testing" those functions...

Also, I found [this discussion](https://github.com/typescript-language-server/typescript-language-server/issues/357) where you can find when single-file mode was introduced to typescript-language-server, down to the exact commit! Hopefully this will eliminate the need for whatever we're trying to ensure ourselves.

**Comment by glepnir** [comment]
oh this is useful issue. so should support it .

**Comment by nickhealy** [comment]
I think this breaks the `denols`, or at least requires some change in documentation. According to the [Deno documention](https://deno.land/manual@v1.29.2/getting_started/setup_your_environment#neovim-06-using-the-built-in-language-server), we should specify separate root directories for our ts and deno projects, so that the tsserver and denols do not both try to attach to the same buffer. With this change introduced, this workaround no longer works, and tsserver still attaches to the buffer in single file mode. When I set `single_file_support` to be `false` in my `lspconfig`, however, the behavior is as expected. 

I agree, though, that we should be able to edit a single typescript file, so I am not sure what the best work around here is, though.

**Comment by anasrar** [comment]
`single_file_support=true` will ignore `root_dir`, this is considered bad for most people that work with `denols` and `tsserver` that use `root_dir` to filter which LSP should be active.

- #2015 
- #2069 
- #2070

**Comment by glepnir** [comment]
```lua
lspconfig.util.on_setup = lspconfig.util.add_hook_before(lspconfig.util.on_setup, function(config)
  local cwd = vim.loop.cwd()
  if config.name == 'tsserver' and vim.fn.filereadable(cwd .. '/deno.json') then
    config.single_file_support = false
  end
end)
```
